File: ios/Classes/observers/ChimeRealtimeDataMessageObserver.swift
//
//  RealtimeDataMessageObserver.swift
//  eggnstone_amazon_chime
//
//  Created by Hendrik Lakämper on 06.02.21.
//

import Foundation
import AmazonChimeSDK
import Flutter

public class ChimeRealtimeDataMessageObserver: DataMessageObserver {
    let _eventSink: FlutterEventSink

    init(eventSink: @escaping FlutterEventSink) {
        self._eventSink = eventSink
    }
    
    public func dataMessageDidReceived(dataMessage: DataMessage) {
        let decoded = String(decoding: dataMessage.data, as: UTF8.self)
        
        let json="""
        {
            "Name": "DataMessageReceived",
            "Arguments": {
                "topic": "\(dataMessage.topic)",
                "data": \(decoded)
            }}
        """
        
        _eventSink(json)
    }
}

File: android/src/main/kotlin/dev/eggnstone/chime/observers/ChimeRealtimeDataMessageObserver.kt
package dev.eggnstone.chime.observers

import com.amazonaws.services.chime.sdk.meetings.realtime.datamessage.DataMessage
import com.amazonaws.services.chime.sdk.meetings.realtime.datamessage.DataMessageObserver
import io.flutter.plugin.common.EventChannel.EventSink
import org.json.JSONObject

class ChimeRealtimeDataMessageObserver(private val _eventSink: EventSink) : DataMessageObserver
{
    private fun convertDataMessageToJson(dataMessage: DataMessage): JSONObject
    {
        val jsonObject = JSONObject()

        jsonObject.put("Topic", dataMessage.topic)
        jsonObject.put("data", String(dataMessage.data, Charsets.UTF_8))

        return jsonObject
    }

    override fun onDataMessageReceived(dataMessage: DataMessage) {
        val jsonObject = JSONObject()
        jsonObject.put("Name", "DataMessageReceived")
        jsonObject.put("Arguments", convertDataMessageToJson(dataMessage))
        _eventSink.success(jsonObject.toString())    }
}
